File: base/src/main/java/me/shetj/base/tools/json/GsonKit.java
package me.shetj.base.tools.json;

import android.support.annotation.Keep;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonParser;
import com.google.gson.JsonPrimitive;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;
import com.google.gson.reflect.TypeToken;

import org.xutils.common.util.LogUtil;

import java.lang.reflect.Type;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Map;

import io.reactivex.annotations.NonNull;

@SuppressWarnings("unchecked")
@Keep
public class GsonKit {
    private static Gson gson = null;
    static {
        if (gson == null) {
            gson = new Gson();
        }
    }
    /**
     * 将对象转换成json格式
     * @param ts
     * @return
     */
    public static String objectToJson(@NonNull Object ts) {
        try {
            String jsonStr = null;
            if (gson != null) {
                jsonStr = gson.toJson(ts);
            }
            return jsonStr;
        }catch (Exception e){
            LogUtil.e(e.getMessage());
            return null;
        }

    }
    /**
     * 将对象转换成json格式(并自定义日期格式)
     *
     * @param ts
     * @return
     */
    public static String objectToJson(@NonNull Object ts, final String dateformat) {
        try {
            gson = new GsonBuilder().registerTypeHierarchyAdapter(Date.class, new JsonSerializer<Date>() {
                @Override
                public JsonElement serialize(Date src, Type typeOfSrc, JsonSerializationContext context) {
                    SimpleDateFormat format = new SimpleDateFormat(dateformat);
                    return new JsonPrimitive(format.format(src));
                }
            }).setDateFormat(dateformat).create();
            return gson.toJson(ts);
        }catch (Exception e){
            LogUtil.e(e.getMessage());
            return null;
        }
    }
    /**
     * 将json格式转换成list对象
     * @param jsonStr
     * @return
     */
    public static <T> List<T> jsonToList(@NonNull String jsonStr) {
       try {
           List<T> objList = null;
           if (gson != null) {
               Type type = new TypeToken<List<T>>() {}.getType();
               objList = gson.fromJson(jsonStr, type);
           }
           return objList;
       }catch (Exception e){
           LogUtil.e(e.getMessage());
           return null;
       }
    }

    /**
     * 转成list中有map的
     *
     * @param gsonString
     * @return
     */
    public static <T> List<Map<String, T>> GsonToListMaps(@NonNull String gsonString) {
        try {
            List<Map<String, T>> list = null;
            if (gson != null) {
                list = gson.fromJson(gsonString,
                        new TypeToken<List<Map<String, T>>>() {
                        }.getType());
            }
            return list;
        }catch (Exception e){
            LogUtil.e(e.getMessage());
            return null;
        }
    }
    /**
     * 转成list
     * 解决泛型问题
     * @param json
     * @param cls
     * @param <T>
     * @return
     */
    public static <T> List<T> jsonToList(@NonNull String json, Class<T> cls) {
        try {
            Gson gson = new Gson();
            List<T> list = new ArrayList<T>();
            JsonArray array = new JsonParser().parse(json).getAsJsonArray();
            for(final JsonElement elem : array){
                list.add(gson.fromJson(elem, cls));
            }
            return list;
        }catch (Exception e){
            LogUtil.e(e.getMessage());
            return null;
        }
    }
    /**
     * 转成map的
     *
     * @param gsonString
     * @return
     */
    public static <T> Map<String, T> jsonToMap(@NonNull String gsonString) {

        try {
            Map<String, T> map = null;
            if (gson != null) {
                map = gson.fromJson(gsonString, new TypeToken<Map<String, T>>() {
                }.getType());
            }
            return map;
        }catch (Exception e){
            LogUtil.e(e.getMessage());
            return null;
        }
    }
    /**
     * 将json转换成bean对象
     * @param jsonStr
     * @return
     */
    public static <T> T jsonToBean(@NonNull String jsonStr, Class<T> cl) {
        try {
            T obj = null;
            if (gson != null) {
                obj = gson.fromJson(jsonStr, cl);
            }
            return obj;
        }catch (Exception e){
            LogUtil.e(e.getMessage());
            return null;
        }
    }
    /**
     * 根据key获得value值
     * @param jsonStr
     * @param key
     * @return
     */
    public static String getJsonValue(@NonNull String jsonStr, String key) {
        try {
            String rusObj = null;
            Map<String, String> rusMap = null;
            if (gson != null) {
                rusMap = gson.fromJson(jsonStr, new TypeToken<Map<String, String>>() {
                }.getType());
            }
            if (rusMap != null && rusMap.size() > 0) {
                rusObj = rusMap.get(key);
            }
            return rusObj;
        }catch (Exception e){
            LogUtil.e(e.getMessage());
            return null;
        }
    }
}

File: base/src/main/java/me/shetj/base/tools/json/GsonKit.kt
package me.shetj.base.tools.json

import androidx.annotation.Keep
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonParser
import com.google.gson.reflect.TypeToken
import io.reactivex.annotations.NonNull
import timber.log.Timber
import java.lang.reflect.Modifier
import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type


/**
 * @author shetj
 */
@Keep
object GsonKit {
    val gson: Gson by lazy {
        GsonBuilder()
                .excludeFieldsWithModifiers(Modifier.FINAL, Modifier.TRANSIENT, Modifier.STATIC)
                .serializeNulls()
                .create()
    }

    /**
     * 将对象转换成json格式
     */
    @JvmStatic
    fun objectToJson(@NonNull ts: Any): String? {
        return try {
            val jsonStr: String? = gson.toJson(ts)
            jsonStr
        } catch (e: Exception) {
            Timber.e(e)
            null
        }

    }

    /**
     * 转成list中有map的
     * TODO need Test
     */
    @JvmStatic
    fun <T> jsonToListMaps(@NonNull gsonString: String): List<Map<String, T>>? {
        return try {
            val list: ArrayList<Map<String, T>> = ArrayList()
            val array = JsonParser().parse(gsonString).asJsonArray
            for (elem in array) {
                list.add(gson.fromJson<Map<String, T>>(gsonString, object : TypeToken<Map<String, Any>>() {
                }.type))
            }
            list
        } catch (e: Exception) {
            Timber.e(e)
            null
        }
    }

    /**
     * 转成list
     * 解决泛型擦除问题
     */
    @JvmStatic
    fun <T> jsonToList(@NonNull json: String, cls: Class<T>): List<T>? {
        return try {
            return gson.fromJson(json, TypeToken.getParameterized(MutableList::class.java, cls).type)
        } catch (e: Exception) {
            Timber.e(e)
            null
        }

    }


    /**
     * 转成list
     * 解决泛型擦除问题
     */
    fun <T> jsonToList2(strJson: String, cls: Class<T>): List<T>? {
        return try {
            val type: Type = ListParameterizedType(cls)
            val list: List<T> = gson.fromJson(strJson, type)
            list
        } catch (e: Exception) {
            Timber.e(e)
            null
        }

    }

    /**
     * 转成map的
     */
    @JvmStatic
    fun jsonToMap(@NonNull gsonString: String): Map<String, Any>? {

        return try {
            val map:Map<String, Any>? = gson.fromJson<Map<String, Any>>(gsonString, object : TypeToken<Map<String, Any>>() {
            }.type)
            map
        } catch (e: Exception) {
            Timber.e(e)
            null
        }

    }

    /**
     * 转成map的
     */
    @JvmStatic
    fun jsonToStringMap(@NonNull gsonString: String): Map<String, String>? {

        return try {
            val map = gson.fromJson<Map<String, String>>(gsonString, object : TypeToken<Map<String, String>>() {

            }.type)
            map
        } catch (e: Exception) {
            Timber.e(e)
            null
        }

    }

    /**
     * 将json转换成bean对象
     */
    @JvmStatic
    fun <T> jsonToBean(@NonNull jsonStr: String, cl: Class<T>): T? {
        return try {
            val obj: T?  = gson.fromJson(jsonStr, cl)
            obj
        } catch (e: Exception) {
            Timber.e(e)
            null
        }

    }

    /**
     * 将json转换成bean对象
     */
    @JvmStatic
    fun <T> jsonToBean(@NonNull jsonStr: String, type: Type): T? {
        return try {
            val obj: T? = gson.fromJson(jsonStr, type)
            obj
        } catch (e: Exception) {
            Timber.e(e)
            null
        }
    }

    /**
     * 根据key获得value值
     */
    @JvmStatic
    fun getJsonValue(@NonNull jsonStr: String, key: String): Any? {
        return try {
            var rusObj: Any? = null
            val rusMap: Map<String, Any>? = gson.fromJson<Map<String, Any>>(jsonStr, object : TypeToken<Map<String, Any>>() {

            }.type)
            if (rusMap != null && rusMap.isNotEmpty()) {
                rusObj = rusMap[key]
            }
            rusObj
        } catch (e: Exception) {
            Timber.e(e)
            null
        }

    }

    private class ListParameterizedType internal constructor(val type: Type) : ParameterizedType {

        override fun getActualTypeArguments(): Array<Type> {
            return arrayOf(type)
        }

        override fun getRawType(): Type {
            return ArrayList::class.java
        }

        override fun getOwnerType(): Type? {
            return null
        }
    }
}